lib/cosmos/dart/script/dart_worker.rb

```diff
@@ -0,0 +1,219 @@
+      end
+
+      if separate_raw_con?(item)
+        values << packet.read_item(item, :RAW)
+        values << packet.read_item(item, :CONVERTED)
+      else
+        values << packet.read_item(item, :RAW)
+      end
+    end
+
+    table_index = 0
+    rows = []
+    values.each_slice(MAX_COLUMNS_PER_TABLE) do |table_values|
+      model = get_decom_table_model(packet_config_id, table_index)
+      row = model.new
+      row.time = ple.time
+      row.ple_id = ple.id
+      row.meta_id = ple.meta_id
+      row.reduced_state = -1
+      table_values.each_with_index do |value, index|
+        item_index = (table_index * MAX_COLUMNS_PER_TABLE) + index
+        row.write_attribute("i#{item_index}", value)
+      end
+      row.save
+      rows << row
+      table_index += 1
+    end
+
+    # Mark ready to reduce
+    rows.each do |row|
+      row.reduced_state = 0
+      row.save
+    end
+  end
+
+  def start
+    begin
+      while true
+        # Check for new data to decom
+        time_start = Time.now
+        PacketLogEntry.where("decom_state = #{PacketLogEntry::NOT_STARTED} and ready = true").where("id % #{@modulus} = #{@mod_index}").order("id ASC").in_batches do |group|
+          group.each do |ple|
+            begin
+              # TODO - Optimize and cache meta and system config and packet config lookup
+
+              # Find meta ple for this ple
+              begin
+                if ple.meta_id != ple.id
+                  meta_ple = PacketLogEntry.find(ple.meta_id)
+                else
+                  meta_ple = ple
+                end
+              rescue => err
+                ple.decom_state = PacketLogEntry::NO_META_PLE
+                ple.save
+                Cosmos::Logger.error("PLE:#{ple.id}:#{ple.decom_state_string}")
+                next
+              end
+
+              # Read the SYSTEM META packet
+              meta_packet = read_packet_from_ple(meta_ple)
+              unless meta_packet
+                ple.decom_state = PacketLogEntry::NO_META_PACKET
+                ple.save
+                Cosmos::Logger.error("PLE:#{ple.id}:#{ple.decom_state_string}")
+                next
+              end
+
+              # Get the System Config
+              system_config_name = meta_packet.read("CONFIG")
+              system_config = SystemConfig.where("name = ?", system_config_name).first
+              begin
+                system_config = SystemConfig.create(:name => system_config_name) unless system_config
+              rescue
+                # Another thread probably already created it - Try to get it one more time
+                system_config = SystemConfig.where("name = ?", system_config_name).first
+              end
+              unless system_config
+                ple.decom_state = PacketLogEntry::NO_SYSTEM_CONFIG
+                ple.save
+                Cosmos::Logger.error("PLE:#{ple.id}:#{ple.decom_state_string}")
```

Comment:
There are a lot of things being logged here which is good. Does this go into the COSMOS server_messages log or some new log? We'll have to create some sort of help to interpret and fix any possible errors which are being logged.

Reply:
I see the message logs under outputs/dart/logs now.

Reply:
Every DART process has its own new log file.